rust_snuba/src/strategies/python.rs

```diff
@@ -11,133 +15,178 @@ use crate::types::BytesInsertBatch;
 
 use crate::config::MessageProcessorConfig;
 
+enum TaskHandle {
+    Procspawn {
+        original_message_meta: Message<()>,
+        join_handle: Mutex<procspawn::JoinHandle<Result<BytesInsertBatch, String>>>
+    },
+    Immediate {
+        original_message_meta: Message<()>,
+        result: Result<BytesInsertBatch, String>
+    },
+}
+
 pub struct PythonTransformStep {
     next_step: Box<dyn ProcessingStrategy<BytesInsertBatch>>,
-    py_process_message: Py<PyAny>,
+    handles: VecDeque<TaskHandle>,
     message_carried_over: Option<Message<BytesInsertBatch>>,
+    processing_pool: Option<procspawn::Pool>,
 }
 
 impl PythonTransformStep {
-    pub fn new<N>(processor_config: MessageProcessorConfig, next_step: N) -> Result<Self, Error>
+    pub fn new<N>(
+        processor_config: MessageProcessorConfig,
+        processes: usize,
+        next_step: N,
+    ) -> Result<Self, Error>
     where
         N: ProcessingStrategy<BytesInsertBatch> + 'static,
     {
         let next_step = Box::new(next_step);
         let python_module = &processor_config.python_module;
         let python_class_name = &processor_config.python_class_name;
-        let code = format!(
-            r#"
-import rapidjson
-from snuba.datasets.processors import DatasetMessageProcessor
-
-from {python_module} import {python_class_name} as Processor
-
-processor = Processor.from_kwargs()
-
-from snuba.consumers.types import KafkaMessageMetadata
-from snuba.processor import InsertBatch
-from datetime import datetime
-from snuba.consumers.consumer import json_row_encoder
-
 
-def _wrapped(message, offset, partition, timestamp):
-    rv = processor.process_message(
-        message=rapidjson.loads(bytearray(message)),
-        metadata=KafkaMessageMetadata(
-            offset=offset,
-            partition=partition,
-            timestamp=timestamp
-        )
-    )
-
-    if rv is None:
-        return rv
-
-    assert isinstance(rv, InsertBatch), "this consumer does not support replacements"
-
-    return [json_row_encoder.encode(row) for row in rv.rows]
-"#
-        );
-
-        let py_process_message = Python::with_gil(|py| -> PyResult<Py<PyAny>> {
-            let fun: Py<PyAny> = PyModule::from_code(py, &code, "", "")?
-                .getattr("_wrapped")?
-                .into();
-            Ok(fun)
-        })?;
+        let processing_pool = if processes <= 1 {
+            Some(procspawn::Pool::builder(processes)
+                .env("RUST_SNUBA_PROCESSOR_MODULE", python_module)
+                .env("RUST_SNUBA_PROCESSOR_CLASSNAME", python_class_name)
```

Comment:
Can we get rid of these eventually? Would be cool if the python strategy could be agnostic of Snuba

Reply:
i think if we move this out of rust_snuba there will have to be a rethinking of the entire approach